File: 课程网站/microlectures/2.3-z-transform/source/first_experiment.py
from pathlib import Path
import numpy as np
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
a=.8;side='right';r=1.2;theta=np.pi/3;N=40;z=r*np.exp(1j*theta)
n=np.arange(N) if side=='right' else -np.arange(1,N+1);terms=(1 if side=='right' else -1)*a**n*z**(-n);s=np.cumsum(terms);inside=r>a if side=='right' else r<a
print('Inside ROC:',inside,'finite sum:',s[-1],'last term modulus:',abs(terms[-1]));print('Algebraic value (not a sum outside ROC):',z/(z-a) if abs(z-a)>1e-12 else 'undefined at pole');assert np.isfinite(s).all()
fig,axes=plt.subplots(1,2,figsize=(10,4));v=np.r_[0,s];axes[0].plot(v.real,v.imag,'.-');axes[0].set(xlabel='Re S_N',ylabel='Im S_N',title='Finite partial sums');axes[0].axis('equal');axes[1].semilogy(np.arange(1,N+1),abs(terms),'.-');axes[1].set(xlabel='Term number',ylabel='Term magnitude');fig.tight_layout();fig.savefig(Path(__file__).with_suffix('.png'));plt.close(fig)

File: 课程网站/microlectures/2.3-z-transform/source/first_experiment.m
% Standalone bilateral z-transform partial sums. Change one parameter at a time.
a=.8; side='right'; r=1.2; theta=pi/3; N=40; z=r*exp(1i*theta);
if strcmp(side,'right'),n=0:N-1;terms=a.^n.*z.^(-n);inside=r>a;else,n=-1:-1:-N;terms=-a.^n.*z.^(-n);inside=r<a;end
S=cumsum(terms);fprintf('Inside ROC: %d; last term modulus %.8g\n',inside,abs(terms(end)));disp('Finite partial sum:');disp(S(end));
if abs(z-a)>1e-12,disp('Algebraic value (not a series sum outside ROC):');disp(z/(z-a));else,disp('Pole: algebraic value undefined');end
assert(all(isfinite(S)));
f=figure('Visible','off');subplot(1,2,1);plot(real([0 S]),imag([0 S]),'.-');axis equal;grid on;xlabel('Re S_N');ylabel('Im S_N');title('Finite partial sums');subplot(1,2,2);semilogy(1:N,abs(terms),'.-');xlabel('Term number');ylabel('Term magnitude');grid on;
exportgraphics(f,fullfile(fileparts(mfilename('fullpath')),'first_experiment_matlab.png'));close(f);
